### PR TITLE
Add password pre-validation service.

### PR DESCRIPTION
**Pull request by dewniMW**
This PR resolves https://github.com/wso2/product-is/issues/6580

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 1]
```diff
@@ -747,4 +749,76 @@ private boolean isFormatCorrect(String regularExpression, String attribute) {
         return m2.matches();
     }
 
+    /**
+     * Pre validate password against the policies defined in the Identity Server during password recovery instance.
+     *
+     * @param confirmationKey Confirmation code for password recovery.
+     * @param password Password to be pre-validated against the policies defined in IS.
+     * @throws IdentityEventException
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Complete the method comment with description

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 2]
```diff
@@ -747,4 +749,76 @@ private boolean isFormatCorrect(String regularExpression, String attribute) {
...
+        String username = user.getUserName();
+
+        preValidatePassword(username, password, userStoreDomain);
+    }
+
+    /**
+     * Pre validate the password against the policies defined in the Identity Server.
+     *
+     * @param username Username
+     * @param password Password to be pre-validated against the policies defined in IS.
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Complete the method comment.
If you are adding `.` at the end of param descriptions, add it for all the descriptions

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 3]
```diff
@@ -747,4 +749,76 @@ private boolean isFormatCorrect(String regularExpression, String attribute) {
...
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Validating password against policies for user: %s in tenant: %s and in user " +
+                    "store: %s", username, tenantDomain, userStoreDomain));
+        }
+
+        Event identityMgtEvent = new Event(eventName, properties);
+        IdentityRecoveryServiceDataHolder.getInstance().getIdentityEventService().handleEvent(identityMgtEvent);
+    }
+
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Shall we remove this new line? WDYT?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 4]
```diff
@@ -747,4 +749,76 @@ private boolean isFormatCorrect(String regularExpression, String attribute) {
...
+        RealmService realmService = IdentityRecoveryServiceDataHolder.getInstance().getRealmService();
+        try {
+            AbstractUserStoreManager userStoreManager = (AbstractUserStoreManager)
+                    realmService.getTenantUserRealm(tenantId).getUserStoreManager();
+            UserStoreManager secondaryUserStoreManager = userStoreManager.getSecondaryUserStoreManager
+                    (userStoreDomain);
+            properties.put(IdentityEventConstants.EventProperty.USER_STORE_MANAGER, secondaryUserStoreManager);
+        } catch (UserStoreException e) {
+            throw Utils.handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED, username,
```
Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: For the Error `IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED` can you append error details?

Comment (somindatommy) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Shall we add the debug log here? WDYT?

Comment (somindatommy) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Don't we have to add the error scenario prior to the error code?

Comment (dewniMW) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: fixed in https://github.com/wso2-extensions/identity-governance/pull/355

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java (hunk 2 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: fixed in https://github.com/wso2-extensions/identity-governance/pull/355

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java (hunk 3 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: fixed in https://github.com/wso2-extensions/identity-governance/pull/355

**Pull request merged**